scylla/src/transport/connection.rs

```diff
@@ -1157,6 +1157,31 @@ impl Connection {
                 .map(|_| ())
         }
 
+        struct Timeouter(Option<tokio::time::Interval>);
+        impl Timeouter {
+            fn new(timeout: Option<Duration>) -> Self {
+                Self(timeout.map(|t| {
+                    let mut interval = tokio::time::interval(t);
+                    interval.reset();
+                    interval
+                }))
+            }
+
+            async fn tick(&mut self) {
+                if let Some(interval) = &mut self.0 {
+                    interval.tick().await;
+                } else {
+                    std::future::pending::<()>().await;
+                }
+            }
+
+            fn reset(&mut self) {
+                if let Some(interval) = &mut self.0 {
+                    interval.reset();
+                }
+            }
+        }
```

Comment:
I don't think there is a need for an ad-hoc abstraction like `Timeouter`. I can see two ways to express it differently and more simply:

- Use an `if` guard in the `select` statement - requires an unwrap:
  ```rust
  tokio::select! {
      () = tokio::time::sleep(keepalive_interval.unwrap()) if keepalive_interval.is_some() => {
          // ...
      },
  }
  ```
- Change `issue_keepalive_query` so that it accepts an optional timeout and enforces it.

Reply:
> * Use an `if` guard in the `select` statement - requires an unwrap:
>       ```rust
>       tokio::select! {
>           () = tokio::time::sleep(keepalive_interval.unwrap()) if keepalive_interval.is_some() => {
>               // ...
>           },
>       }
>       ```

I'm afraid this isn't semantically equivalent to the current code.
- in your suggestion, the sleep is interrupted and dropped once another `select!` branch completes, and this happens after every `keepalive_interval`. 
- in my version, the state (how much time has already elapsed _since sending the first keepalive that hasn't been followed by any response_ - sorry for the complicated phrase here; I dunno how to put it into easier words) is saved between iterations. 
Do you agree @piodul?

Reply:
OK, I see why my suggestion is incorrect, but I think the first bullet point is easy to fix: introduce a mutable variable that keeps the time of the last successful keepalive, and then use `tokio::time::sleep_until(last_successful_keepalive_time + keepalive_timeout)`.

Reply:
I've realised that both your suggestion and the current implementation are semantically different that what is desired (and what I stated above). The difference is that you suggest to count time since _last successful keepalive response_ (and this is done by the code), whereas what should be done is, most probably, to count time since _last keepalive not yet having faced any keepalive response_.

The code can be fixed by swapping lines, from:
```rust
timeouter.reset();
// Sending keepalives finished before receiving new tick.
// Wait for the new tick and start over.
interval.tick().await;

// implicit continue
```
to:
```rust
// Sending keepalives finished before receiving new tick.
// Wait for the new tick and start over.
interval.tick().await;
timeouter.reset();

// implicit continue
```

To illustrate the difference:
```
                     <-----------------------------> timeout len
           <-----------------------------> timeout len
    a      b         c                   d         e = d + (c - b)
----|------|---------|-------------------|---------|------------------>
a - keepalive 1 sent
b - received response to keepalive 1
c - keepalive 2 sent (no response arrived)
d - timeout by your suggestion and previous code
e - timeout by my description and updated code
```

Reply:
I'm still in favor of `Timeouter`. Such abstractions, even if "ad-hoc", enable delegating details, while leaving only meaningful logic.

Reply:
> ```
>                      <-----------------------------> timeout len
>            <-----------------------------> timeout len
>     a      b         c                   d         e = d + (c - b)
> ----|------|---------|-------------------|---------|------------------>
> a - keepalive 1 sent
> b - received response to keepalive 1
> c - keepalive 2 sent (no response arrived)
> d - timeout by your suggestion and previous code
> e - timeout by my description and updated code
> ```

What are the arguments in favor of your suggestion?

I think about it like this: a successful response gives us a reason to think that the connection works right now, therefore we should count the idle timeout starting from the last response. Glancing at the CPP driver implementation this [seems to be the case](https://github.com/scylladb/cpp-driver/blob/eb6b70d138752c5e6a487eeaf69e6baa4d6652a4/src/connection.cpp#L237), and the docstring seems to agree with it (emphasis mine):

* Sets the amount of time a connection is allowed to be _without a successful heartbeat response_ before being terminated and scheduled for reconnection.

I'd vote for this interpretation as it will make it easier for us to have the same behavior in cpp wrapper over the rust driver.

Reply:
> I'm still in favor of `Timeouter`. Such abstractions, even if "ad-hoc", enable delegating details, while leaving only meaningful logic.

I am still not. I'm not convinced that this particular abstraction improves the readability rather than obfuscating the code. The amount of complexity it hides is tiny compared to the size of the type definition and overhead related to reading and understanding it. I think in this particular case it helps to be explicit about the details, and by using names such as `last_successful_keepalive_time` instead of non-descript `timeouter` it will actually _improve_ readability.

Reply:
I refactored the code to remove `Timeouter` altogether.